README.md
*FFTL* — generalised FFTLog for Python
======================================

The *FFTL* package for Python contains a routine to calculate integral
transforms of the type *ã(k) = ∫ a(r) T(kr) dr* for arbitrary kernels *T*.  It
uses a modified FFTLog [2] method of Hamilton [1] to efficiently compute the
transform on logarithmic input and output grids.

The package supports any array implementation such as Numpy, JAX, Torch, dask,
etc. as long as it provides a reasonable set of standard Array API functions
(e.g. `log()`, `exp()`, `linspace()`, etc.) and Fast Fourier Transforms
(`fft.rfft()` and `fft.irfft()`).

Besides the generalised FFTLog algorithm, the package also provides a number of
standard integral transforms implemented using SciPy.


Installation
------------

Install with pip:

    pip install fftl

For development, it is recommended to clone the GitHub repository, and perform
an editable pip installation.

The core package only requires `numpy`.

The standard integral transforms in the `fftl.scipy` module additionally
require `scipy`.  You can install *FFTL* with these additional dependencies
using `pip install fftl[scipy]`.


Usage
-----

The core functionality of the package is provided by the [`fftl`] module.  The
[`fftl()`] routine computes the generalised FFTLog integral transform for a
given kernel.

For convenience, a number of standard integral transforms are implemented in
the [`fftl.scipy`] module.

[`fftl`]: https://fftl.readthedocs.io/en/latest/fftl.html
[`fftl()`]: https://fftl.readthedocs.io/en/latest/fftl.html#fftl.fftl
[`fftl.scipy`]: https://fftl.readthedocs.io/en/latest/scipy.html


User manual
-----------

* [`fftl` — Generalised FFTLog][`fftl`]
* [`fftl.scipy` — Standard Integral Transforms using SciPy][`fftl.scipy`]


References
----------

1.  Hamilton A. J. S., 2000, MNRAS, 312, 257 (astro-ph/9905191)
2.  Talman J. D., 1978, J. Comp. Phys., 29, 35


CHANGELOG.md
Changelog
=========

Unreleased
----------
* support for Array API
* refactored standard transforms

v2022.7.18
----------
* use flit for packaging
* add Stieltjes transform
* add Laplace transform
* add Hankel transform

v2022.7.15
----------
* module for standard integral transforms

v2022.7.13
----------
* initial version


fftl.py
# author: Nicolas Tessore <n.tessore@ucl.ac.uk>
# license: MIT
'''generalised FFTLog module'''

__version__ = '2022.7.13'

import numpy as np

PI = np.pi


def fftl(u, r, ar, q=0.0, args=(), kr=1.0, krgood=True, deriv=False):
    r'''generalised FFTLog for integral transforms

    Computes integral transforms for arbitrary kernels using a generalisation
    of Hamilton's method [1]_ for the FFTLog algorithm [2]_.

    The kernel of the integral transform is characterised by the coefficient
    function ``u``, see notes below, which must be callable and accept complex
    input arrays.  Additional arguments for ``u`` can be passed with the
    optional ``args`` parameter.

    The function to be transformed must be given on a logarithmic grid ``r``.
    The result of the integral transform is similarly computed on a logarithmic
    grid ``k = kr/r``, where ``kr`` is a scalar constant (default: 1) which
    shifts the logarithmic output grid.  The selected value of ``kr`` is
    automatically changed to the nearest low-ringing value if ``krgood`` is
    true (the default).

    The integral transform can optionally be biased, see notes below.

    The function can optionally at the same time return the derivative of the
    integral transform with respect to the logarithm of ``k``, by setting
    ``deriv`` to true.

    Parameters
    ----------
    u : callable
        Coefficient function.  Must have signature ``u(x, *args)`` and support
        complex input arrays.
    r : array_like (N,)
        Grid of input points.  Must have logarithmic spacing.
    ar : array_like (..., N)
        Function values.  If multidimensional, the integral transform applies
        to the last axis, which must agree with input grid.
    q : float, optional
        Bias parameter for integral transform.
    args : tuple, optional
        Additional arguments for the coefficient function ``u``.
    kr : float, optional
        Shift parameter for logarithmic output grid.
    krgood : bool, optional
        Change given ``kr`` to the nearest value fulfilling the low-ringing
        condition.
    deriv : bool, optional
        Also return the first derivative of the integral transform.

    Returns
    -------
    k : array_like (N,)
        Grid of output points.
    ak : array_like (..., N)
        Integral transform evaluated at ``k``.
    dak : array_like (..., N), optional
        If ``deriv`` is true, the derivative of ``ak`` with respect to the
        logarithm of ``k``.

    Notes
    -----
    Computes integral transforms of the form

    .. math::

        \tilde{a}(k) = \int_{0}^{\infty} \! a(r) \, T(kr) \, dr

    for arbitrary kernels :math:`T`.

    If :math:`a(r)` is given on a logarithmic grid of :math:`r` values, the
    integral transform can be computed for a logarithmic grid of :math:`k`
    values with a modification of Hamilton's FFTLog algorithm,

    .. math::

        U(x) = \int_{0}^{\infty} \! t^x \, T(t) \, dt \;.

    The generalised FFTLog algorithm therefore only requires the coefficient
    function :math:`U` for the given kernel.  Everything else, and in
    particular how to construct a well-defined transform, remains exactly the
    same as in Hamilton's original algorithm.

    The transform can optionally be biased,

    .. math::

        \tilde{a}(k) = k^{-q} \int_{0}^{\infty} \! [a(r) \, r^{-q}] \,
                                                    [T(kr) \, (kr)^q] \, dr \;,

    where :math:`q` is the bias parameter.  The respective biasing factors
    :math:`r^{-q}` and :math:`k^{-q}` for the input and output values are
    applied internally.

    References
    ----------
    .. [1] Hamilton A. J. S., 2000, MNRAS, 312, 257 (astro-ph/9905191)
    .. [2] Talman J. D., 1978, J. Comp. Phys., 29, 35

    Examples
    --------
    Compute the one-sided Laplace transform of the hyperbolic tangent function.
    The kernel of the Laplace transform is :math:`\exp(-kt)`, which determines
    the coefficient function.

    >>> from scipy.special import gamma, digamma
    >>>
    >>> def u_laplace(x):
    ...     # requires Re(x) = q > -1
    ...     return gamma(1 + x)

    Create the input function values on a logarithmic grid.

    >>> r = np.logspace(-4, 4, 100)
    >>> ar = np.tanh(r)
    >>>
    >>> import matplotlib.pyplot as plt
    >>> plt.loglog(r, ar)
    >>> plt.xlabel('$r$')
    >>> plt.ylabel('$\\tanh(r)$')
    >>> plt.show()

    Compute the Laplace transform, and compare with the analytical result.

    >>> from fftl import fftl
    >>>
    >>> k, ak = fftl(u_laplace, r, ar)
    >>>
    >>> lt = (digamma((k+2)/4) - digamma(k/4) - 2/k)/2
    >>>
    >>> plt.loglog(k, ak)
    >>> plt.loglog(k, lt, ':')
    >>> plt.xlabel('$k$')
    >>> plt.ylabel('$L[\\tanh](k)$')
    >>> plt.show()

    The numerical Laplace transform has an issue on the right, which is due to
    the circular nature of the FFTLog integral transform.  The effect is
    mitigated by computing a biased transform with ``q = 0.5``.  Good values of
    the bias parameter ``q`` depend on the shape of the input function.

    >>> k, ak = fftl(u_laplace, r, ar, q=0.5)
    >>>
    >>> plt.loglog(k, ak)
    >>> plt.loglog(k, lt, ':')
    >>> plt.xlabel('$k$')
    >>> plt.ylabel('$L[\\tanh](k)$')
    >>> plt.show()

    '''

    if np.ndim(r) != 1:
        raise TypeError('r must be 1d array')
    if np.shape(ar)[-1] != len(r):
        raise TypeError('last axis of ar must agree with r')

    # inputs
    n = len(r)
    lnkr = np.log(kr)

    # make sure given r is logarithmic grid
    if not np.allclose(r, np.geomspace(r[0], r[-1], n)):
        raise ValueError('r it not a logarithmic grid')

    # log spacing
    dlnr = (np.log(r[-1]) - np.log(r[0]))/(n-1)

    # low-ringing condition
    if krgood:
        y = PI/dlnr
        um = np.exp(-1j*y*lnkr)*u(q + 1j*y, *args)
        a = np.angle(um)/PI
        lnkr = lnkr + dlnr*(a - np.round(a))

    # transform factor
    y = np.linspace(0, 2*PI*(n//2)/(n*dlnr), n//2+1)
    um = np.exp(-1j*y*lnkr)*u(q + 1j*y, *args)

    # low-ringing kr should make last coefficient real
    if krgood and not np.isclose(um[-1].imag, 0):
        raise ValueError('unable to construct low-ringing transform, '
                         'try odd number of points or different q')

    # fix last coefficient to real when n is even
    if not n & 1:
        um.imag[-1] = 0

    # bias input
    if q != 0:
        ar = ar*r**(-q)

    # set up k in log space
    k = np.exp(lnkr)/r[::-1]

    # transform via real FFT
    cm = np.fft.rfft(ar, axis=-1)
    cm *= um
    ak = np.fft.irfft(cm, n, axis=-1)
    ak[..., :] = ak[..., ::-1]

    # debias output
    ak /= k**(1+q)

    # output grid and transform
    result = (k, ak)

    # derivative
    if deriv:
        cm *= -(1 + q + 1j*y)
        dak = np.fft.irfft(cm, n, axis=-1)
        dak[..., :] = dak[..., ::-1]
        dak /= k**(1+q)
        result = result + (dak,)

    # return chosen outputs
    return result


fftl/jax.py
# author: Nicolas Tessore <n.tessore@ucl.ac.uk>
# license: MIT
'''
:mod:`fftl.jax` --- FFTL for JAX
================================

.. currentmodule:: fftl.jax

The :mod:`fftl.jax` module provides JAX implementations for a number of
standard integral transforms.

Interface
---------

.. autofunction:: fftl

Standard Integral Transforms
----------------------------

.. autoclass:: HankelTransform
.. autoclass:: LaplaceTransform
.. autoclass:: SphericalHankelTransform
.. autoclass:: StieltjesTransform

'''

from dataclasses import dataclass
import jax.numpy as jnp
from jax import custom_jvp, jit, lax
from . import newfftl

SR_PI = jnp.pi**0.5
LN_PI = jnp.log(jnp.pi)
LN_SR_2PI = jnp.log(2*jnp.pi)/2


def digamma(z):
    '''Compute digamma(z) following Kölbig (1972).'''

    X0 = 7.

    # these are C[k] = b[k]/b[k-1], k = 1, 2, 3, ...,
    # where b[k] = B[2k]/2k, k > 0, with B[2k] the Bernoulli coefficient
    # and b[0] was set to unity
    DIGAM_C = (
        0.083333333333333333333333333333333333333333333333333,
        -0.10000000000000000000000000000000000000000000000000,
        -0.47619047619047619047619047619047619047619047619048,
        -1.0500000000000000000000000000000000000000000000000,
        -1.8181818181818181818181818181818181818181818181818,
        -2.7842490842490842490842490842490842490842490842491,
        -3.9507959479015918958031837916063675832127351664255,
        -5.3191176470588235294117647058823529411764705882353,
        -6.8897614970981966020017327439878841883992997884999,
        -8.6629363965872047117630357879292157907062049127839,
        -10.638716670322826914883842419410301607075404632359,
        -12.817127438511138934655019255253540747149387392426,
        # -15.198176866891341798615255817348329785001055849892,
        # -17.781867530496201807977033961060480139438162444348,
        # -20.568200221926880643403873557351656580971638924982,
        # -23.557175180388933362770162012497160914155672721662,
        # -26.748792476906204901749058388489603740929195010834,
        # -30.143052132279567979802139252370977832315962353185,
        # -33.739954152530241976514641568519088474590403161705,
        # -37.539498539383718370168841295418657387101825078851,
    )

    # map left half-plane to right half-plane, keep track of correction
    # to improve accuracy and handle the case of negative integers correctly,
    # use the fact that tan(pi*z) has unit period in x

    ncor = jnp.where(z.real < 0,
                     1/z + jnp.pi/jnp.tan(jnp.pi*(z - jnp.ceil(z.real))),
                     0)
    z = jnp.where(z.real < 0, -z, z)

    # range reduction: map z to x >= X0, keep track of correction

    rcor = jnp.zeros_like(z)
    for _ in range(int(X0)):
        u = (z.real < X0)
        rcor = rcor + u/z
        z = z + u

    # compute digamma for x >= X0

    r = 1/(z*z)
    s = jnp.zeros_like(z)
    t = jnp.ones_like(z)
    for c in DIGAM_C:
        t = c*r*t
        s = s + t

    log_z = jnp.log(z)

    return log_z - 1/(2*z) - s - rcor - ncor


def _log_sin_pi(z):
    '''Compute ln sin(pi*z).'''
    x = z.real
    pi_xi = jnp.pi * (x - jnp.floor(x))
    if jnp.iscomplexobj(z):
        conj = (z.imag < 0)
        pi_y = jnp.pi * jnp.where(conj, -z.imag, z.imag)
        u = pi_y + jnp.log(jnp.exp(-2*pi_y) * jnp.sin(pi_xi)**2
                           + jnp.expm1(-2*pi_y)**2/4)/2
        v = jnp.arctan(jnp.tanh(pi_y)/jnp.tan(pi_xi)) - jnp.floor(x)*jnp.pi
        v = jnp.where(conj, -v, v)
    else:
        u = jnp.log(jnp.abs(jnp.sin(pi_xi)))
        v = -jnp.pi*jnp.floor(x)
    return lax.complex(u, v)


@custom_jvp
def loggamma(z):
    '''Compute ln Gamma(z) following Kölbig (1972).'''

    X0 = 7.

    # these are C[k] = b[k]/b[k-1], k = 1, 2, 3, ...,
    # where b[k] = B[2k]/2k/(2k-1), k > 0, with B[2k] the Bernoulli coefficient
    # and b[0] was set to unity
    GAMMA_C = (
        0.083333333333333333333333333333333333333333333333333,
        -0.033333333333333333333333333333333333333333333333333,
        -0.28571428571428571428571428571428571428571428571429,
        -0.75000000000000000000000000000000000000000000000000,
        -1.4141414141414141414141414141414141414141414141414,
        -2.2780219780219780219780219780219780219780219780220,
        -3.3429811866859623733719247467438494934876989869754,
        -4.6099019607843137254901960784313725490196078431373,
        -6.0792013209689970017662347741069566368229115780882,
        -7.7510483548411831631564004418314036022108149219645,
        -9.6255055588635100658472859985140824064015565721345,
        -11.702594617771039896858930624361928508266831966998,
        # -13.982322717540034454726035351960463402200971381901,
        # -16.464692157866853525904661075056000129109409670693,
        # -19.149703654897440599031192622361887161594284516363,
        # -22.037357426815453790978538656852182790661758352522,
        # -25.127653538911889453158206364944779271781971070784,
        # -28.420592010435021238099159866521207670469335933003,
        # -31.916172846988066734540877159409948557044975963775,
        # -35.614396050184553325544798152063854444173526356859,
    )

    # map left half-plane to right half-plane, keep track of correction

    ncor = jnp.where(z.real < 0, _log_sin_pi(z) - LN_PI, 0)
    nfac = jnp.where(z.real < 0, -1, +1)
    z = jnp.where(z.real < 0, 1 - z, z)

    # range reduction: map z to x >= X0, keep track of correction

    rcor = jnp.zeros_like(z)
    if jnp.iscomplexobj(z):
        for _ in range(int(X0)):
            u = (z.real < X0)
            rcor = rcor + u*lax.complex(jnp.log(jnp.abs(z)), jnp.angle(z))
            z = z + u
    else:
        for _ in range(int(X0)):
            u = (z < X0)
            rcor = rcor + u*jnp.log(z)
            z = z + u

    # compute loggamma for x >= X0

    r = 1/(z*z)
    s = jnp.zeros_like(z)
    t = jnp.ones_like(z)
    for c in GAMMA_C:
        t = c*r*t
        s = s + t

    log_z = jnp.log(z)

    return nfac*((z - 0.5)*log_z - z + z*s + LN_SR_2PI - rcor) - ncor


@loggamma.defjvp
def loggamma_jvp(primals, tangents):
    z, = primals
    t, = tangents
    return loggamma(z), digamma(z) * t


def poch(z, m):
    '''Pochhammer symbol for complex arguments'''
    return jnp.exp(loggamma(z+m) - loggamma(z))


def beta(a, b):
    '''Beta function for complex arguments'''
    return jnp.exp(loggamma(a) + loggamma(b) - loggamma(a+b))


fftl = jit(newfftl(jnp), static_argnames=('u', 'deriv'))


@dataclass(frozen=True)
class HankelTransform:
    r'''Hankel transform on a logarithmic grid.

    The Hankel transform is here defined as

    .. math::

        \tilde{a}(k) = \int_{0}^{\infty} \! a(r) \, J_\mu(kr) \, r \, dr \;,

    where :math:`J_\mu` is the Bessel function of order :math:`\mu`.  The order
    can in general be any real or complex number.  The transform is orthogonal
    and normalised: applied twice, the original function is returned.

    The Hankel transform is equivalent to a normalised spherical Hankel
    transform  with the order and bias shifted by one half.  Special cases are
    :math:`\mu = 1/2`, which is related to the Fourier sine transform,

    .. math::

        \tilde{a}(k)
        = \sqrt{\frac{2}{\pi}} \int_{0}^{\infty} \! a(r) \,
                                    \frac{\sin(kr)}{\sqrt{kr}} \, r \, dr \;,

    and :math:`\mu = -1/2`, which is related to the Fourier cosine transform,

    .. math::

        \tilde{a}(k)
        = \sqrt{\frac{2}{\pi}} \int_{0}^{\infty} \! a(r) \,
                                    \frac{\cos(kr)}{\sqrt{kr}} \, r \, dr \;.

    See Also
    --------
    SphericalHankelTransform :
        Equivalent transform with order and bias shifted by one half.

    Examples
    --------
    Compute the Hankel transform for parameter ``mu = 1``.

    >>> import jax.numpy as jnp
    >>>
    >>> # some test function
    >>> p, q = 2.0, 0.5
    >>> r = jnp.logspace(-2, 2, 1000)
    >>> ar = r**p*jnp.exp(-q*r)
    >>>
    >>> # compute a biased transform
    >>> from fftl.jax import HankelTransform
    >>> hankel = HankelTransform(1.0)
    >>> k, ak = hankel.fftl(r, ar, q=0.1)

    Compare with the analytical result.

    >>> from scipy.special import gamma, hyp2f1
    >>> res = k*q**(-3-p)*gamma(3+p)*hyp2f1((3+p)/2, (4+p)/2, 2, -(k/q)**2)/2
    >>>
    >>> import matplotlib.pyplot as plt
    >>> plt.plot(k, ak, '-k', label='numerical')            # doctest: +SKIP
    >>> plt.plot(k, res, ':r', label='analytical')          # doctest: +SKIP
    >>> plt.xscale('log')                                   # doctest: +SKIP
    >>> plt.yscale('symlog', linthresh=1e0,
    ...            subs=[2, 3, 4, 5, 6, 7, 8, 9])           # doctest: +SKIP
    >>> plt.ylim(-5e-1, 1e2)                                # doctest: +SKIP
    >>> plt.legend()                                        # doctest: +SKIP
    >>> plt.show()

    '''

    mu: complex

    def __call__(self, x):
        return 2**x*poch((1+self.mu-x)/2, x)

    @fftl.wrap
    def fftl(self, r, ar, **kwargs):
        return fftl(self, r, ar*r, **kwargs)


@dataclass(frozen=True)
class LaplaceTransform:
    r'''Laplace transform on a logarithmic grid.

    The Laplace transform is defined as

    .. math::

        \tilde{a}(k) = \int_{0}^{\infty} \! a(r) \, e^{-kr} \, dr \;.

    Examples
    --------
    Compute the Laplace transform.

    >>> import jax.numpy as jnp
    >>>
    >>> # some test function
    >>> p, q = 2.0, 0.5
    >>> r = jnp.logspace(-2, 2, 1000)
    >>> ar = r**p*jnp.exp(-q*r)
    >>>
    >>> # compute a biased transform
    >>> from fftl.jax import LaplaceTransform
    >>> laplace = LaplaceTransform()
    >>> k, ak = laplace.fftl(r, ar, q=0.7)

    Compare with the analytical result.

    >>> from scipy.special import gamma
    >>> res = gamma(p+1)/(q + k)**(p+1)
    >>>
    >>> import matplotlib.pyplot as plt
    >>> plt.loglog(k, ak, '-k', label='numerical')          # doctest: +SKIP
    >>> plt.loglog(k, res, ':r', label='analytical')        # doctest: +SKIP
    >>> plt.legend()                                        # doctest: +SKIP
    >>> plt.show()

    '''

    def __call__(self, x):
        return jnp.exp(loggamma(1+x))

    fftl = fftl


@dataclass(frozen=True)
class SphericalHankelTransform:
    r'''Hankel transform with spherical Bessel functions.

    The spherical Hankel transform is here defined as

    .. math::

        \tilde{a}(k) = \int_{0}^{\infty} \! a(r) \, j_\mu(kr) \, r^2 \, dr \;,

    where :math:`j_\mu` is the spherical Bessel function of order :math:`\mu`.
    The order can in general be any real or complex number.  The transform is
    orthogonal, but unnormalised: applied twice, the original function is
    multiplied by :math:`\pi/2`.

    The spherical Hankel transform is equivalent to an unnormalised Hankel
    transform with the order and bias shifted by one half.  Special cases are
    :math:`\mu = 0`, which is related to the Fourier sine transform,

    .. math::

        \tilde{a}(k)
        = \int_{0}^{\infty} \! a(r) \, \frac{\sin(kr)}{kr} \, r^2 \, dr \;,

    and :math:`\mu = -1`, which is related to the Fourier cosine transform,

    .. math::

        \tilde{a}(k)
        = \int_{0}^{\infty} \! a(r) \, \frac{\cos(kr)}{kr} \, r^2 \, dr \;.

    See Also
    --------
    HankelTransform :
        Equivalent transform with order and bias shifted by one half.

    Examples
    --------
    Compute the spherical Hankel transform for parameter ``mu = 1``.

    >>> import jax.numpy as jnp
    >>>
    >>> # some test function
    >>> p, q = 2.0, 0.5
    >>> r = jnp.logspace(-2, 2, 1000)
    >>> ar = r**p*jnp.exp(-q*r)
    >>>
    >>> # compute a biased transform
    >>> from fftl.jax import SphericalHankelTransform
    >>> sph_hankel = SphericalHankelTransform(1.0)
    >>> k, ak = sph_hankel.fftl(r, ar, q=0.1)

    Compare with the analytical result.

    >>> from scipy.special import gamma
    >>> u = (1 + k**2/q**2)**(-p/2)*q**(-p)*gamma(1+p)/(k**2*(k**2 + q**2)**2)
    >>> v = k*(k**2*(2 + p) - p*q**2)*jnp.cos(p*jnp.arctan(k/q))
    >>> w = q*(k**2*(3 + 2*p) + q**2)*jnp.sin(p*jnp.arctan(k/q))
    >>> res = u*(v + w)
    >>>
    >>> import matplotlib.pyplot as plt
    >>> plt.plot(k, ak, '-k', label='numerical')            # doctest: +SKIP
    >>> plt.plot(k, res, ':r', label='analytical')          # doctest: +SKIP
    >>> plt.xscale('log')                                   # doctest: +SKIP
    >>> plt.yscale('symlog', linthresh=1e0,
    ...            subs=[2, 3, 4, 5, 6, 7, 8, 9])           # doctest: +SKIP
    >>> plt.ylim(-1e0, 1e3)                                 # doctest: +SKIP
    >>> plt.legend()                                        # doctest: +SKIP
    >>> plt.show()

    '''

    mu: complex

    def __call__(self, x):
        return 2**(x-1)*SR_PI*poch((2+self.mu-x)/2, (2*x-1)/2)

    @fftl.wrap
    def fftl(self, r, ar, **kwargs):
        return fftl(self, r, ar*r**2, **kwargs)


@dataclass(frozen=True)
class StieltjesTransform:
    r'''Generalised Stieltjes transform on a logarithmic grid.

    The generalised Stieltjes transform is defined as

    .. math::

        \tilde{a}(k) = \int_{0}^{\infty} \! \frac{a(r)}{(k + r)^\rho} \, dr \;,

    where :math:`\rho` is a positive real number.

    The integral can be computed as a :func:`fftl` transform in :math:`k' =
    k^{-1}` if it is rewritten in the form

    .. math::

        \tilde{a}(k) = k^{-\rho} \int_{0}^{\infty} \! a(r) \,
                                        \frac{1}{(1 + k'r)^\rho} \, dr \;.

    Warnings
    --------
    The Stieltjes FFTLog transform is often numerically difficult.

    Examples
    --------
    Compute the generalised Stieltjes transform with ``rho = 2``.

    >>> import jax.numpy as jnp
    >>>
    >>> # some test function
    >>> s = 0.1
    >>> r = jnp.logspace(-4, 2, 100)
    >>> ar = r/(s + r)**2
    >>>
    >>> # compute a biased transform with shift
    >>> from fftl.jax import StieltjesTransform
    >>> stieltjes = StieltjesTransform(2.)
    >>> k, ak = stieltjes.fftl(r, ar, kr=1e-2)

    Compare with the analytical result.

    >>> res = (2*(s-k) + (k+s)*jnp.log(k/s))/(k-s)**3
    >>>
    >>> import matplotlib.pyplot as plt
    >>> plt.loglog(k, ak, '-k', label='numerical')          # doctest: +SKIP
    >>> plt.loglog(k, res, ':r', label='analytical')        # doctest: +SKIP
    >>> plt.legend()                                        # doctest: +SKIP
    >>> plt.show()

    Compute the derivative in two ways and compare with numerical and
    analytical results.

    >>> # compute Stieltjes transform with derivative
    >>> k, ak, akp = stieltjes.fftl(r, ar, kr=1e-1, deriv=True)
    >>>
    >>> # derivative by rho+1 transform
    >>> stieltjes_d = StieltjesTransform(stieltjes.rho+1)
    >>> k_, takp = stieltjes_d.fftl(r, ar, kr=1e-1)
    >>> takp *= -stieltjes.rho*k_
    >>>
    >>> # analytical derivative
    >>> aakp = -((-5*k**2+4*k*s+s**2+2*k*(k+2*s)*jnp.log(k/s))/(k-s)**4)
    >>>
    >>> # show
    >>> plt.loglog(k, -akp, '-k', label='deriv')            # doctest: +SKIP
    >>> plt.loglog(k_, -takp, '-.b', label='rho+1')         # doctest: +SKIP
    >>> plt.loglog(k, -aakp, ':r', label='analytical')      # doctest: +SKIP
    >>> plt.legend()                                        # doctest: +SKIP
    >>> plt.show()

    '''

    rho: float

    def __call__(self, x):
        return beta(1+x, -1-x+self.rho)

    @fftl.wrap
    def fftl(self, r, ar, *, kr, **kwargs):
        kr = r[-1]*r[0]/kr

        k, ak, *dak = fftl(self, r, ar, kr=kr, **kwargs)

        k, ak = 1/k[::-1], ak[::-1]
        ak = ak * k**(-self.rho)
        if dak:
            dak[0] = -dak[0][::-1] * k**(-self.rho) - self.rho*ak

        return (k, ak, *dak)


fftl/_core.py
# author: Nicolas Tessore <n.tessore@ucl.ac.uk>
# license: MIT
'''
Core Functionality (:mod:`fftl`)
================================

.. currentmodule:: fftl

The main functionality of the package is provided by the :func:`fftl` routine
to compute the generalised FFTLog integral transform for a given kernel.


List of functions
-----------------

.. autosummary::
   :toctree: reference
   :nosignatures:

   fftl

'''

import numpy as np

PI = np.pi


def fftl(u, r, ar, q=0.0, args=(), kr=1.0, krgood=True, deriv=False):
    r'''generalised FFTLog for integral transforms

    Computes integral transforms for arbitrary kernels using a generalisation
    of Hamilton's method [1]_ for the FFTLog algorithm [2]_.

    The kernel of the integral transform is characterised by the coefficient
    function ``u``, see notes below, which must be callable and accept complex
    input arrays.  Additional arguments for ``u`` can be passed with the
    optional ``args`` parameter.

    The function to be transformed must be given on a logarithmic grid ``r``.
    The result of the integral transform is similarly computed on a logarithmic
    grid ``k = kr/r``, where ``kr`` is a scalar constant (default: 1) which
    shifts the logarithmic output grid.  The selected value of ``kr`` is
    automatically changed to the nearest low-ringing value if ``krgood`` is
    true (the default).

    The integral transform can optionally be biased, see notes below.

    The function can optionally at the same time return the derivative of the
    integral transform with respect to the logarithm of ``k``, by setting
    ``deriv`` to true.

    Parameters
    ----------
    u : callable
        Coefficient function.  Must have signature ``u(x, *args)`` and support
        complex input arrays.
    r : array_like (N,)
        Grid of input points.  Must have logarithmic spacing.
    ar : array_like (..., N)
        Function values.  If multidimensional, the integral transform applies
        to the last axis, which must agree with input grid.
    q : float, optional
        Bias parameter for integral transform.
    args : tuple, optional
        Additional arguments for the coefficient function ``u``.
    kr : float, optional
        Shift parameter for logarithmic output grid.
    krgood : bool, optional
        Change given ``kr`` to the nearest value fulfilling the low-ringing
        condition.
    deriv : bool, optional
        Also return the first derivative of the integral transform.

    Returns
    -------
    k : array_like (N,)
        Grid of output points.
    ak : array_like (..., N)
        Integral transform evaluated at ``k``.
    dak : array_like (..., N), optional
        If ``deriv`` is true, the derivative of ``ak`` with respect to the
        logarithm of ``k``.

    Notes
    -----
    Computes integral transforms of the form

    .. math::

        \tilde{a}(k) = \int_{0}^{\infty} \! a(r) \, T(kr) \, dr

    for arbitrary kernels :math:`T`.

    If :math:`a(r)` is given on a logarithmic grid of :math:`r` values, the
    integral transform can be computed for a logarithmic grid of :math:`k`
    values with a modification of Hamilton's FFTLog algorithm,

    .. math::

        U(x) = \int_{0}^{\infty} \! t^x \, T(t) \, dt \;.

    The generalised FFTLog algorithm therefore only requires the coefficient
    function :math:`U` for the given kernel.  Everything else, and in
    particular how to construct a well-defined transform, remains exactly the
    same as in Hamilton's original algorithm.

    The transform can optionally be biased,

    .. math::

        \tilde{a}(k) = k^{-q} \int_{0}^{\infty} \! [a(r) \, r^{-q}] \,
                                                    [T(kr) \, (kr)^q] \, dr \;,

    where :math:`q` is the bias parameter.  The respective biasing factors
    :math:`r^{-q}` and :math:`k^{-q}` for the input and output values are
    applied internally.

    References
    ----------
    .. [1] Hamilton A. J. S., 2000, MNRAS, 312, 257 (astro-ph/9905191)
    .. [2] Talman J. D., 1978, J. Comp. Phys., 29, 35

    Examples
    --------
    Compute the one-sided Laplace transform of the hyperbolic tangent function.
    The kernel of the Laplace transform is :math:`\exp(-kt)`, which determines
    the coefficient function.

    >>> from scipy.special import gamma, digamma
    >>>
    >>> def u_laplace(x):
    ...     # requires Re(x) = q > -1
    ...     return gamma(1 + x)

    Create the input function values on a logarithmic grid.

    >>> r = np.logspace(-4, 4, 100)
    >>> ar = np.tanh(r)
    >>>
    >>> import matplotlib.pyplot as plt
    >>> plt.loglog(r, ar)
    >>> plt.xlabel('$r$')
    >>> plt.ylabel('$\\tanh(r)$')
    >>> plt.show()

    Compute the Laplace transform, and compare with the analytical result.

    >>> from fftl import fftl
    >>>
    >>> k, ak = fftl(u_laplace, r, ar)
    >>>
    >>> lt = (digamma((k+2)/4) - digamma(k/4) - 2/k)/2
    >>>
    >>> plt.loglog(k, ak)
    >>> plt.loglog(k, lt, ':')
    >>> plt.xlabel('$k$')
    >>> plt.ylabel('$L[\\tanh](k)$')
    >>> plt.show()

    The numerical Laplace transform has an issue on the right, which is due to
    the circular nature of the FFTLog integral transform.  The effect is
    mitigated by computing a biased transform with ``q = 0.5``.  Good values of
    the bias parameter ``q`` depend on the shape of the input function.

    >>> k, ak = fftl(u_laplace, r, ar, q=0.5)
    >>>
    >>> plt.loglog(k, ak)
    >>> plt.loglog(k, lt, ':')
    >>> plt.xlabel('$k$')
    >>> plt.ylabel('$L[\\tanh](k)$')
    >>> plt.show()

    '''

    if np.ndim(r) != 1:
        raise TypeError('r must be 1d array')
    if np.shape(ar)[-1] != len(r):
        raise TypeError('last axis of ar must agree with r')

    # inputs
    n = len(r)
    lnkr = np.log(kr)

    # make sure given r is logarithmic grid
    if not np.allclose(r, np.geomspace(r[0], r[-1], n)):
        raise ValueError('r it not a logarithmic grid')

    # log spacing
    dlnr = (np.log(r[-1]) - np.log(r[0]))/(n-1)

    # low-ringing condition
    if krgood:
        y = PI/dlnr
        um = np.exp(-1j*y*lnkr)*u(q + 1j*y, *args)
        a = np.angle(um)/PI
        lnkr = lnkr + dlnr*(a - np.round(a))

    # transform factor
    y = np.linspace(0, 2*PI*(n//2)/(n*dlnr), n//2+1)
    um = np.exp(-1j*y*lnkr)*u(q + 1j*y, *args)

    # low-ringing kr should make last coefficient real
    if krgood and not np.isclose(um[-1].imag, 0):
        raise ValueError('unable to construct low-ringing transform, '
                         'try odd number of points or different q')

    # fix last coefficient to real when n is even
    if not n & 1:
        um.imag[-1] = 0

    # bias input
    if q != 0:
        ar = ar*r**(-q)

    # set up k in log space
    k = np.exp(lnkr)/r[::-1]

    # transform via real FFT
    cm = np.fft.rfft(ar, axis=-1)
    cm *= um
    ak = np.fft.irfft(cm, n, axis=-1)
    ak[..., :] = ak[..., ::-1]

    # debias output
    ak /= k**(1+q)

    # output grid and transform
    result = (k, ak)

    # derivative
    if deriv:
        cm *= -(1 + q + 1j*y)
        dak = np.fft.irfft(cm, n, axis=-1)
        dak[..., :] = dak[..., ::-1]
        dak /= k**(1+q)
        result = result + (dak,)

    # return chosen outputs
    return result
